cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp

```diff
@@ -0,0 +1,272 @@
+    const std::int32_t sg_count = wg_size / sg_size;
+    const std::int32_t sg_id = sg.get_group_id();
+
+    const std::int32_t sg_local_id = sg.get_local_id();
+
+    const std::int32_t int_max = dal::detail::limits<std::int32_t>::max();
+
+    Float x = objective_func[local_id];
+    std::int32_t x_index = local_id;
+
+    Float res_max = sycl::reduce_over_group(sg, x, maximum<Float>());
+
+    std::int32_t res_index =
+        sycl::reduce_over_group(sg, res_max == x ? x_index : int_max, minimum<std::int32_t>());
+
+    if (sg_local_id == 0) {
+        sg_cache_values[sg_id] = res_max;
+        sg_cache_index[sg_id] = res_index;
+    }
+
+    item.barrier(sycl::access::fence_space::local_space);
+
+    if (sg_id == 0 && sg_local_id < sg_count) {
+        x = sg_cache_values[sg_local_id];
+        x_index = sg_cache_index[sg_local_id];
+        res_max = sycl::reduce_over_group(sg, x, maximum<Float>());
+        res_index =
+            sycl::reduce_over_group(sg, res_max == x ? x_index : int_max, minimum<std::int32_t>());
+
+        for (std::int32_t group_index = sg_size; group_index < sg_count; group_index += sg_size) {
+            x = sg_cache_values[group_index + sg_local_id];
+            x_index = sg_cache_index[group_index + sg_local_id];
+
+            const Float inner_max = sycl::reduce_over_group(sg, x, maximum<Float>());
+            if (inner_max > res_max) {
+                res_max = inner_max;
+                res_index = sycl::reduce_over_group(sg,
+                                                    res_max == x ? x_index : int_max,
+                                                    minimum<std::int32_t>());
+            }
+        }
+
+        if (sg_local_id == 0) {
+            sg_cache_values[max_val_ind] = res_max;
+            sg_cache_index[max_val_ind] = res_index;
+        }
+    }
+
+    item.barrier(sycl::access::fence_space::local_space);
+}
+
+template <typename Float>
+sycl::event solve_smo(sycl::queue& q,
+                      const pr::ndview<Float, 2>& kernel_values,
+                      const pr::ndview<std::int32_t, 1>& ws_indices,
+                      const pr::ndarray<Float, 1>& labels,
+                      const pr::ndview<Float, 1>& grad,
+                      const std::int64_t row_count,
+                      const std::int64_t ws_count,
+                      const std::int64_t max_inner_iter,
+                      const Float C,
+                      const Float eps,
+                      const Float tau,
+                      pr::ndview<Float, 1>& alpha,
+                      pr::ndview<Float, 1>& delta_alpha,
+                      pr::ndview<Float, 1>& grad_diff,
+                      pr::ndview<std::int32_t, 1>& inner_iter_count,
+                      const dal::backend::event_vector& deps = {}) {
+    ONEDAL_PROFILER_TASK(solve_smo, q);
+    ONEDAL_ASSERT(row_count > 0);
+    ONEDAL_ASSERT(row_count <= dal::detail::limits<std::int32_t>::max());
```

Comment:
If you don't allow using values higher than dal::detail::limits<std::int32_t>::max() then you need to check it at the algo params initialization block  and throw an exception if required, smth like
    if (data.get_row_count() > de::limits<Index>::max()) {
        throw domain_error(msg::invalid_range_of_rows());
    }

Reply:
ASSERT isn't enough here due to it will not prevent passing big row_count in release